HIP-clang support

psychocoderHPC opened the pull request:
Add support for the compiler HIP-clang.

- extent CMake `ALPAKA_HIP_PLATFORM` with the option `clang`
- add define `BOOST_COMP_HIP` if hip-clang is used

I know that we currently not cover hip-clang in the CI. The reason is that hip-clang is currently not final released and the installation is not as easy. Never the less we need these changes to work more closely together with AMD and Cray. At all HIP-clang is nearly equal to HIP-hcc but is more relaxed with the function attributes `__host__` and `__device__` so that #802 is not needed.
If someone breaks HIP-clang support somehow with a follow up PR it is not dramatic I will keep an eye on it and fix it if needed.

BenjaminW3 commented on include/alpaka/kernel/TaskKernelGpuHipRt.hpp:
```diff
@@ -76,14 +76,16 @@ namespace alpaka
-#pragma clang diagnostic pop
-
+#   pragma clang diagnostic pop
+#endif
```
Is the `pragma` unsupported or the `static_assert`? I would like to keep the `static_assert` or at least add a comment why this does not work.

psychocoderHPC replied:
The evaluation of the static assert expression failed but I now got it solved.

```
                    static_assert(
                        std::is_same<
                            decltype(kernelFnObj(
                                alpaka::core::declval<acc::AccGpuHipRt<TDim, TIdx> const>(),
                                args...)),
                        void>::value,
                        "The TKernelFnObj is required to return void!");
```

sbastrakov commented on include/alpaka/core/Utility.hpp:
```diff
@@ -27,7 +27,7 @@ namespace alpaka
         // This function can be used only within std::decltype().
         //-----------------------------------------------------------------------------
-#if BOOST_LANG_CUDA && BOOST_COMP_CLANG_CUDA
+#if BOOST_LANG_CUDA && BOOST_COMP_CLANG_CUDA || BOOST_COMP_HIP
```
If the (lack of) brackets what was intended here?

psychocoderHPC replied:
yes they are not needed

sbastrakov commented on include/alpaka/queue/QueueHipRtBlocking.hpp:
```diff
@@ -106,7 +106,7 @@ namespace alpaka
                     hipStream_t m_HipQueue;
 
-#if BOOST_COMP_HCC  // NOTE: workaround for unwanted nonblocking hip streams for HCC (NVCC streams are blocking)
+#if BOOST_COMP_HCC || BOOST_COMP_HIP  // NOTE: workaround for unwanted nonblocking hip streams for HCC (NVCC streams are blocking)
```
In the comment should probably also be `HCC or HIP`?

sbastrakov replied:
Same for other occurrences of this comment (next files).

psychocoderHPC replied:
I will remove this for HIP-clang and will add a `static_assert` that we currently not support callbacks with `HIP-clang`.
This line is only a workaround because the hip callbacks are not synchronous. The current workaround adds polling but this is destroying all performance e.g. for PIConGPU.

sbastrakov commented on test/unit/core/src/BoostPredefTest.cpp:
```diff
@@ -31,9 +31,12 @@ TEST_CASE("printDefines", "[core]")
     std::cout << "BOOST_COMP_NVCC:" << BOOST_COMP_NVCC << std::endl;
 #endif
-#if BOOST_COMP_HCC
+#if BOOST_COMP_HCC || BOOST_COMP_HIP
```
Is it intended that for HIP now both lines with `BOOST_COMP_HCC:` and `BOOST_COMP_HIP:` will be printed?

psychocoderHPC replied:
thx must be removed

psychocoderHPC committed:
HIP-clang support

Add support for the compiler HIP-clang.

- extent CMake `ALPAKA_HIP_PLATFORM` with the option `clang`
- add define `BOOST_COMP_HIP` if hip-clang is usedA
- disable callback functions for HIP-clang (workaround for non blocking
hip callback is to slow)

psychocoderHPC commented:
ready

sbastrakov reviewed (approved):
Looks good to me, thanks!
